### ISSUE TITLE
Quota and replication

### ISSUE THREAD

**Issue by kargakis** [kubernetes/kubernetes#25146]
The quota system, in its current form, doesn't play well with high-level resources such as replication controllers/replica sets/deployments. While quota will correctly not allow my replicaset to scale up to the N pods I have asked for -because N pods may exceed {cpu,memory,pods} quota- the replicaset is blocked in limbo as its desired state has been updated but may never be fulfilled. This has implications for deployments too. A deployment that manages such a replicaset will never complete successfully because the size of the replicaset is not reflecting the reality. It's probably a problem for the user to solve but do we want to do something about it? Can we do something about it?

If this is a dupe of discussion elsewhere, please link it and close this.

@derekwaynecarr @kubernetes/rh-platform-management @bgrant0607

**Comment by bgrant0607** [comment]
cc @davidopp @mml

**Comment by bgrant0607** [comment]
Related: #21912

**Comment by bgrant0607** [comment]
I don't think there's a strict dupe of this.

I totally agree that it makes sense to check pod quota when creating or scaling a controller.  

I think the tricky part is how to avoid double-charging a pod. 

Maybe controllerRef could help. The controller could squat on pod quota up front and then pods could steal it back from the controller indicated by controllerRef as they are created. Deployment would have 2 levels of squatting (Deployment to ReplicaSet to Pods).

**Comment by davidopp** [comment]
Yes, I can't think of anything better than something like what Brian suggested.

**Comment by derekwaynecarr** [comment]
I am not sure I agree with the original problem description.  Do you have
an opinion on Job and concurrency?  Asking the quota system to understand
all potential higher order resources that may result in a pod creation is
non-trivial.  There were previous requests to not charge for quota on
compute resources until point of scheduling, but that also has pros and
cons.

**Comment by derekwaynecarr** [comment]
I think it's fair to warn the user that a replica may not be satisfied due
to quota usage unless something else frees it up.  In some ways, events do
that now but maybe we can make it more prominent.  There are a number of
reasons why a replica set it deployment could not be satisfied though.
Lack of capacity, etc. separate from quota.  I need to think on this more.
Other areas of complication is where server side defaulting occurs for CPU
and memory only at the point of pod creation.  That can make reserving
quota with the replica size change difficult since you need to simulate a
pod creation and any auto sizing, limit range, stuff.  For example, we have
a cluster resource overcommit admission controller in OpenShift that would
make doing this difficult unless it's done at point of pod creation.

**Comment by davidopp** [comment]
I think it's pretty reasonable for someone to want an admission decision at the time they create or update whatever user-facing abstractions we tell them to use to manage their workloads (Deployment, ReplicaSet, etc.). The way we do it today seems a lot less user-friendly than something that comes in earlier in the submission process.

Note that what Brian's suggesting (assuming I understand it correctly) would still allow you to do quota exactly as today (i.e. on pods). But you would also be able to do it on things like Job, Deployment, etc. as well.

I think quota checking at admission time would provide value even without a full simulation to see if the pods would go pending due to insufficient free resources, unsatisfiable label selectors, etc. After all, today's quota checking doesn't check any of those other things, and I think people still find it useful.

Your point about LimitRange, initial resources, etc. is an interesting one. Arguably we should be applying those to the PodTemplate, not to pods individually.

Regarding "Asking the quota system to understand all potential higher order resources that may result in a pod creation" -- that's a reasonable point, and maybe for that reason each quota-able controller should have its own quota admission controller? (Some, like DaemonSet, can't sensibly have quota.)

I think the "charge quota only for running pods" idea you referred to (#13567) is orthogonal.

**Comment by ironcladlou** [comment]
The theme of the current discussion seems to be new mechanisms for quota violation prediction.

I would also suggest the value of the existing mechanism could be extended by a richer failure context attached to the things that are failing. For example, if the deployment controller is trying to scale up a replicationController and knows progress is stalled by pod create failures resulting from count quota violation, that information could be propagated all the way back to the deployment through status.

The current failures are accurate and consistent and technically the system state is correct, but the failure are presented in a context far away from the user (who is trying to monitor a deployment, not the pods created by an RC owned by the deployment). The user is then burdened with the task of locating the error and reconstructing a context, which the system could have done for them. 

The responsibility to provide that context would be in the controller managing those high level object (the deployment controller could provide the reason it thinks progress has stopped, e.g. as part of https://github.com/kubernetes/kubernetes/pull/19343). The replicaset manager could add such context to the RC itself which would be useful to users attempting to manage replicas without deployments. And of course, if a user's just trying to create a pod directly, the API failure they receive immediately is fine for that context.

Does it make any sense to approach this from a UX perspective?

**Comment by ironcladlou** [comment]
As @bgrant0607, https://github.com/kubernetes/kubernetes/issues/21912 could go a long way towards this. It's not clear to me yet whether events or status are considered more discoverable/meaningful to users for these scenarios.

**Comment by derekwaynecarr** [comment]
@deads2k - fyi

**Comment by derekwaynecarr** [comment]
@ironcladlou - I agree with roll-up on failure conditions as a good first step before doing anything more radical.

**Comment by kargakis** [comment]
> As @bgrant0607 mentioned, #21912 could go a long way towards this. It's not clear to me yet whether events or status are considered more discoverable/meaningful to users for these scenarios.

I would lean towards surfacing to status. Events is another area we need to work on. Related: https://github.com/kubernetes/kubernetes/issues/11994

**Comment by soltysh** [comment]
@kargakis the problem with status is that you need to then parse the events and know which to present and which not. Which basically means you're tightly coupling the controller logic (responsible for updating the status) with specific events. I'd prefer having an easier way to collect events with regards to specific objects, exactly like in the attached issue, you've pointed to.

**Comment by derekwaynecarr** [comment]
@soltysh - if every object had the equivalent of NodeConditions that explained potential reasons for why a desired state could not be achieved that would potentially provide a cleaner method of introspection.

**Comment by kargakis** [comment]
@soltysh isn't each controller responsible for emitting those events anyway? Agreed with @derekwaynecarr. Conditions is the way to go.

**Comment by derekwaynecarr** [comment]
Per the reservation discussion, I have a general proposal on supporting reservations here though it is primarily motivated around addressing latency between the creation of a resource and the quota recalculation:  https://github.com/kubernetes/kubernetes/pull/20113

I will need to think if it can be extended to support controllers making a reservation on behalf of a not yet created resource.  controllerRef is a potential aid there.

**Comment by soltysh** [comment]
:+1: for the conditions

**Comment by bgrant0607** [comment]
@derekwaynecarr @soltysh @kargakis I am in favor of conditions that are common across all resources. I've mentioned this a bunch of places, such as:
https://github.com/kubernetes/kubernetes/issues/1899#issuecomment-93879097
https://github.com/kubernetes/kubernetes/issues/7856#issuecomment-99711909

This is needed in order to facilitate higher-level orchestration on top of the K8s APIs. The original /operations API was intended to be used for that, but was modeled after an imperative API (GCE's), which wasn't the right thing.

I could imagine introducing a polymorphic /conditions subresource or somesuch.

**Comment by kargakis** [comment]
@derekwaynecarr not sure which sig owns quota, can you relabel this issue accordignly?

**Comment by xiangpengzhao** [comment]
/sig api-machinery
maybe this sig...

**Comment by davidopp** [comment]
Actually it's
/sig scheduling

**Comment by xiangpengzhao** [comment]
:+1:

**Comment by sftim** [comment]
/retitle Warn or error when a workload object (eg Deployment) would breach a quota

**Comment by sftim** [comment]
I like the idea of:

- a warning if you try to scale a Deployment so that the Deployment on its own breaches a quota that is enforced
- (harder)
  a warning if you try to scale a Deployment so that the namespace overall is in breach of an enforced quota

Related: https://kubernetes.io/blog/2022/12/20/validating-admission-policies-alpha/

perhaps a future enhancement to ValidatingAdmissionPolicies will have a threshold for _warning_ the client.
